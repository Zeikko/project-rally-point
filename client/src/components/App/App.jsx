import PropTypes from 'prop-types'
import React from 'react'
import { connect } from 'react-redux'
import TopBar from '../TopBar/TopBar'
import * as propTypes from '../../constants/prop-types'
import PlayersList from '../PlayersList/PlayersList'
import Game from '../Game/Game'

function App(props) {
  const {
    dispatch,
    user,
    game,
    players,
  } = props
  return (
    <div>
      <TopBar user={user} dispatch={dispatch} />
      <Game players={players} game={game} user={user} dispatch={dispatch} />
      {game.data && <PlayersList players={players} game={game} dispatch={dispatch} />}
    </div>
  )
}

/* istanbul ignore next */
const mapStateToProps = ({
  user,
  game,
  players,
}) => ({
  user,
  game,
  players,
})

App.propTypes = {
  dispatch: PropTypes.func.isRequired,
  user: propTypes.user.isRequired,
  game: propTypes.game.isRequired,
  players: propTypes.players.isRequired,
}

export { App as AppWithoutConnect }
export default connect(mapStateToProps)(App)
