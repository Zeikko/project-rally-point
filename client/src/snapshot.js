import renderer from 'react-test-renderer'
import 'jest-styled-components'

export function matchComponentToSnapshot(component) {
  const json = renderer.create(component).toJSON()
  expect(json).toMatchSnapshot()
}