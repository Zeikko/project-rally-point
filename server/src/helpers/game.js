import _ from 'lodash'
import db from '../db'
import gameStatuses from '../../../common/game-statuses.json'
import playerRoles from '../../../common/player-roles.json'

export async function shouldStartCaptainVote(gameId) {
  const game = await db('game')
    .first('*')
    .where({ id: gameId })
  const numberOfPlayers = await db('player')
    .count('id')
    .where({ gameId })
  return parseInt(numberOfPlayers[0].count, 10) === game.maxPlayers
}

export async function shouldStartSquadLeaderPick(gameId) {
  const game = await db('game')
    .first('*')
    .where({ id: gameId })
  const numberOfCaptainVotes = await db('captainVote')
    .count('id')
    .where({ gameId })
  return parseInt(numberOfCaptainVotes[0].count, 10) === game.maxPlayers
}

export async function shouldStartSquadMemberPick(gameId) {
  const game = await db('game')
    .first('*')
    .where({ id: gameId })
  const numberOfSquadLeaders = await db('player')
    .select('id')
    .where({
      gameId,
      role: playerRoles.SQUAD_LEADER,
      team: game.teamWithTurnToPick,
    })
  return (numberOfSquadLeaders.length + 1) === getOptimalNumberOfSquads(game)
}

export async function shouldStartPlayingGame(gameId) {
  const game = await db('game')
    .first('*')
    .where({ id: gameId })
  const numberOfPlayers = await db('player')
    .count('id')
    .where({ gameId })
    .whereNotNull('team')
  return parseInt(numberOfPlayers[0].count, 10) === game.maxPlayers
}

export async function startCaptainVote(gameId) {
  const games = await db('game')
    .update({ status: gameStatuses.VOTE_CAPTAINS })
    .where({ id: gameId })
    .returning('*')
  return _.first(games)
}

export async function startSquadLeaderPick(gameId) {
  const games = await db('game')
    .update({
      status: gameStatuses.SQUAD_LEADER_PICK,
      teamWithTurnToPick: 1,
    })
    .where({ id: gameId })
    .returning('*')
  return _.first(games)
}

export async function startSquadMemberPick(gameId) {
  const games = await db('game')
    .update({ status: gameStatuses.SQUAD_MEMBER_PICK })
    .where({ id: gameId })
    .returning('*')
  return _.first(games)
}

export async function startPlayingGame(gameId) {
  const games = await db('game')
    .update({ status: gameStatuses.PLAYING })
    .where({ id: gameId })
    .returning('*')
  return _.first(games)
}

export function getGameStatus(gameId) {
  return db('game')
    .first('status')
    .where({ id: gameId })
}

export async function passSquadLeaderPickTurn(game) {
  const games = await db('game')
    .update({
      teamWithTurnToPick: (game.teamWithTurnToPick % 2) + 1,
    })
    .where({ id: game.id })
    .returning('*')
  return _.first(games)
}

export async function passPlayerPickTurn(game, playersCount) {
  const nextTeamWithTurnToPick = await calculatePickTurn(game, playersCount)
  const games = await db('game')
    .update({
      teamWithTurnToPick: nextTeamWithTurnToPick,
    })
    .where({ id: game.id })
    .returning('*')
  return _.first(games)
}

async function calculatePickTurn(game, playersCount) {
  const numberOfSquads = getOptimalNumberOfSquads(game)
  if (playersCount % numberOfSquads === 0) {
    return (game.teamWithTurnToPick % 2) + 1
  }
  return game.teamWithTurnToPick
}

export function getOptimalNumberOfSquads(game) {
  return Math.ceil(game.maxPlayers / 2 / 9)
}
