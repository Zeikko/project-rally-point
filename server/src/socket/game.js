import _ from 'lodash'
import db from '../db'
import actions from '../../../common/actions.json'

export default function handleGame(io, socket, action) {
  switch (action.type) {
    case actions.GET_GAME_REQUEST:
      return getGameRequest(io, socket, action)
    default:
      return null
  }
}

export async function getGameRequest(io, socket) {
  try {
    const game = await db('game')
      .first('*')
      .where({
        status: 'queue',
      })
    socket.emit('action', { type: actions.GET_GAME_SUCCESS, data: game })
  } catch (error) {
    socket.emit('action', { type: actions.GET_GAME_ERROR })
  }
}

export async function shouldStartCaptainVote(gameId) {
  const game = await db('game')
    .first('*')
    .where({ id: gameId })
  const numberOfPlayers = await db('player')
    .count('id')
    .where({ gameId })
  return parseInt(numberOfPlayers[0].count, 10) === game.maxPlayers
}

export async function startCaptainVote(gameId) {
  const games = await db('game')
    .update({ status: 'voting captains' })
    .where({ id: gameId })
    .returning('*')
  return _.first(games)
}

export function createNewGame() {
  return db('game').insert({})
}

export function getGameStatus(gameId) {
  return db('game')
    .first('status')
    .where({ id: gameId })
}

