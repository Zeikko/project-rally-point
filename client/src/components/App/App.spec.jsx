import React from 'react'
import { AppWithoutConnect } from './App'
import { matchComponentToSnapshot } from '../../test/snapshot'
import initialState from '../../initial-state'
import game from '../../../../fixtures/game.json'

describe('App', () => {
  it('renders initial state correctly', () => {
    matchComponentToSnapshot(<AppWithoutConnect
      dispatch={() => {}}
      {...initialState}
    />)
  })

  it('renders correctly with game', () => {
    matchComponentToSnapshot(<AppWithoutConnect
      dispatch={() => {}}
      {...initialState}
      gameState={{ game }}
    />)
  })
})
