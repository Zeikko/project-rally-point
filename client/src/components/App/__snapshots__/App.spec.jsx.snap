// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly with game 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  background-color: #EEE;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 10px;
}

.c1 {
  font-size: 20px;
}

.c2 {
  font-size: 18px;
  border: solid black 1px;
  background-color: #AAA;
  cursor: pointer;
}

<div>
  <div
    className="c0"
  >
    <div
      className="c1"
    >
      Project Rally Point
    </div>
    <a
      href="/api/auth/login"
    >
      <img
        alt="Login"
        src="/img/steam_login.png"
      />
    </a>
  </div>
  <div>
    <div>
      QUEUE
    </div>
    <div>
      Players: 
      0
       / 
      48
    </div>
    <button
      className="c2"
      onClick={[Function]}
    >
      Join Game
    </button>
  </div>
</div>
`;

exports[`App renders initial state correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  background-color: #EEE;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 10px;
}

.c1 {
  font-size: 20px;
}

<div>
  <div
    className="c0"
  >
    <div
      className="c1"
    >
      Project Rally Point
    </div>
    <a
      href="/api/auth/login"
    >
      <img
        alt="Login"
        src="/img/steam_login.png"
      />
    </a>
  </div>
</div>
`;
